src/main/java/io/vertx/proton/impl/ProtonReceiverImpl.java

```diff
@@ -231,6 +249,53 @@ void onDelivery() {
     }
   }
 
+  private boolean checkMaxMessageSize(final UnsignedLong maxMessageSize, final Delivery delivery, final Receiver receiver) {
+    if(maxMessageSizeExceeded) {
+      // Already previously exceeded.
+      // Drop any accumulated payload for the delivery, wont be processed.
+      receiver.recv();
+
+      return true;
+    }
+
+    // Check size, using any available now in the delivery, and any previously buffered partial payload
+    long payloadLength = delivery.available();
+    if(splitContent != null) {
+      payloadLength += splitContent.remaining();
+    }
+
+    long max = maxMessageSize.longValue();
+    if(max > 0 && payloadLength > max) {
+      maxMessageSizeExceeded = true;
+
+      // Drop any accumulated payload for the delivery, wont be processed
+      splitContent = null;
+      receiver.recv();
+
+      handleMaxMessageSizeExceeded(maxMessageSize, receiver);
+
+      return true;
+    }
+
+    return false;
+  }
+
+  private void handleMaxMessageSizeExceeded(final UnsignedLong maxMessageSize, final Receiver receiver) {
+    try {
+      LOG.debug("delivery received exceeding max-message-size of " + maxMessageSize + " bytes");
+      if (maxMessageSizeExceededHandler != null) {
+        maxMessageSizeExceededHandler.handle(this);
+      }
+    } finally {
+      // Detach link if handler has not sent a detach frame already
+      if (!getReceiver().detached() && isOpen()) {
```

Comment:
`!receiver.detached() && isOpen()` ?

Reply:
yep, meant to do that and missed it, updated now.